java-client/session/src/main/java/io/deephaven/client/impl/ClientChannelFactoryDefaulter.java

```diff
@@ -0,0 +1,72 @@
+//
+// Copyright (c) 2016-2024 Deephaven Data Labs and Patent Pending
+//
+package io.deephaven.client.impl;
+
+import io.deephaven.annotations.BuildableStyle;
+import io.deephaven.ssl.config.SSLConfig;
+import io.deephaven.uri.DeephavenTarget;
+import io.grpc.ManagedChannel;
+import org.immutables.value.Value.Default;
+import org.immutables.value.Value.Immutable;
+
+import java.util.Optional;
+
+/**
+ * A client channel factory that may update {@link ClientConfig} values before handing off to a {@link #delegate()}.
+ */
+@Immutable
+@BuildableStyle
+public abstract class ClientChannelFactoryDefaulter implements ClientChannelFactory {
+
+    public static Builder builder() {
+        return ImmutableClientChannelFactoryDefaulter.builder();
+    }
+
+    /**
+     * The delegated client channel factory. By default, is {@link ClientChannelFactory#defaultInstance()}.
+     */
+    @Default
+    public ClientChannelFactory delegate() {
+        return ClientChannelFactory.defaultInstance();
+    }
+
+    /**
+     * The ssl config. This will set {@link ClientConfig#ssl() ssl} for the {@code config} in
+     * {@link #create(ClientConfig)} if it has not already been set and the {@link ClientConfig#target() target} is
+     * {@link DeephavenTarget#isSecure() secure}.
+     */
+    public abstract Optional<SSLConfig> ssl();
+
+    /**
+     * The user-agent. This will set {@link ClientConfig#userAgent()} for the {@code config} in
+     * {@link #create(ClientConfig)} if it has not already been set.
+     */
+    public abstract Optional<String> userAgent();
+
+    /**
+     * Creates a managed channel. Will update {@code config} with the defaults as specified by {@link #ssl()} and
+     * {@link #userAgent()} before delegating to {@link #delegate()}.
+     */
+    @Override
+    public final ManagedChannel create(ClientConfig config) {
+        if (ssl().isPresent() && !config.ssl().isPresent() && config.target().isSecure()) {
+            config = config.withSsl(ssl().get());
+        }
+        if (userAgent().isPresent() && !config.userAgent().isPresent()) {
+            config = config.withUserAgent(userAgent().get());
+        }
+        return delegate().create(config);
+    }
+
+    public interface Builder {
```

Comment:
personal opinion: it is far more important to javadoc the builder side than the getter side - we, the DHC developers are consuming the properties, but customers are trying to build them to pass to us.

Reply:
It's a fair point, although I think taken as a whole, documenting the resulting object propertiies is "easier" to reason about IMO, both from developer and user perspective (assuming they know how to make the leap from builder to object); I've added basic verbiage that points from the builder methods to the relevant object properties. If we were redesigning our usage of immutables, it would be worthwhile to re-examine our hiding of the implementations; if we made the code-generated stuff public, those builders come with boilerplate javadoc (essentially, the same as what I copied onto these methhods).

Reply:
I personally struggle with the indirection between builder interface and immutable model, and needing to go through the generated impl to bridge the gap to find out what the expected values are, the defaults, the docs to go with them. It doesn't happen often that I have to make a sample app using these, but pretty much without exception I have to fight these docs every time.